File: c/medium/minimum_removals_to_balance_array.c
#include <stdio.h>
#include <stdlib.h>

int cmp(const void *a, const void *b) {
    return (*(int *)a - *(int *)b);
}

int min_removal(int *nums, int n, int k) {
    qsort(nums, n, sizeof(int), cmp);

    if ((long long)nums[0] * k >= nums[n - 1]) {
        return 0;
    }

    int min_removals = n;
    int left = 0;

    for (int right = 0; right < n; right++) {
        while (left < right && (long long)nums[left] * k < nums[right]) {
            left++;
        }

        int window_size = right - left + 1;
        int removals = n - window_size;
        if (removals < min_removals) {
            min_removals = removals;
        }
    }

    return min_removals;
}

File: rust/medium/minimum_removals_to_balance_array.rs
fn min_removal(mut nums: Vec<i32>, k: i32) -> i32 {
    nums.sort();
    let n = nums.len();

    if (nums[0] as i64) * (k as i64) >= nums[n - 1] as i64 {
        return 0;
    }

    let mut min_removals = n as i32;
    let mut left = 0usize;

    for right in 0..n {
        while left < right && (nums[left] as i64) * (k as i64) < nums[right] as i64 {
            left += 1;
        }

        let window_size = (right - left + 1) as i32;
        min_removals = min_removals.min(n as i32 - window_size);
    }

    min_removals
}
